### ISSUE TITLE
ERROR: Unauthorized

### ISSUE THREAD

**Issue by armab** [StackStorm/hubot-stackstorm#33]
Running it on production with authentication enabled. At start it works ok, but after some period of time when trying to run chatops command I get:

![error](http://i.imgur.com/g8jvROd.png)

Version: 
`hubot-stackstorm` `0.2.1`

Log: 
```log
[Wed Jul 29 2015 16:50:44 GMT+0000 (UTC)] INFO Connecting...
[Wed Jul 29 2015 16:50:44 GMT+0000 (UTC)] INFO Logged in as chatopsbot of ..., but not yet connected
[Wed Jul 29 2015 16:50:45 GMT+0000 (UTC)] INFO Slack client now connected
[Wed Jul 29 2015 16:50:45 GMT+0000 (UTC)] INFO Performing authentication...
[Wed Jul 29 2015 16:50:45 GMT+0000 (UTC)] INFO /opt/hubot/node_modules/hubot-stackstorm/scripts/stackstorm.js is using deprecated documentation syntax
[Wed Jul 29 2015 16:50:45 GMT+0000 (UTC)] INFO hubot-redis-brain: Using default redis on localhost:6379
[Wed Jul 29 2015 16:50:45 GMT+0000 (UTC)] INFO Loading commands....
[Wed Jul 29 2015 16:50:45 GMT+0000 (UTC)] INFO hubot-redis-brain: Data for hubot brain retrieved from Redis
[Wed Jul 29 2015 16:52:45 GMT+0000 (UTC)] INFO Loading commands....
[Wed Jul 29 2015 16:54:45 GMT+0000 (UTC)] INFO Loading commands....
...
(note exactly 24h interval)
...
[Thu Jul 30 2015 16:48:47 GMT+0000 (UTC)] INFO Loading commands....
[Thu Jul 30 2015 16:50:47 GMT+0000 (UTC)] INFO Loading commands....
[Thu Jul 30 2015 16:50:47 GMT+0000 (UTC)] ERROR Failed to retrieve commands: {
    "faultstring": "Unauthorized"
}
[Thu Jul 30 2015 16:52:47 GMT+0000 (UTC)] INFO Loading commands....
[Thu Jul 30 2015 16:52:47 GMT+0000 (UTC)] ERROR Failed to retrieve commands: {
    "faultstring": "Unauthorized"
}
```

Seems something related to expired tokens,
because after restarting hubot it returns back to normal.

**Comment by pixelrebel** [comment]
I receive this error from the get-go of a hubot start.  
```
[Thu Aug 27 2015 16:03:04 GMT-0700 (PDT)] INFO Loading commands....
[Thu Aug 27 2015 16:03:04 GMT-0700 (PDT)] ERROR Failed to retrieve commands from "http://localhost:9101": {
    "faultstring": "Unauthorized"
}
```

**Comment by jfryman** [comment]
@pixelrebel Is the username/password that you've passed to Hubot is valid and works via either the CLI or WebUI?

**Comment by pixelrebel** [comment]
Oops.  I didn't do that step, in the past I had auth disabled.  How to I pass the credentials to hubot? Can I set an environment variable in [hubot.conf](https://github.com/StackStorm/showcase-ansible-chatops/blob/master/hubot/hubot.conf)?

**Comment by pixelrebel** [comment]
@jfryman Oops. I didn't do that step, in the past I had auth disabled. How to I pass the credentials to hubot? Can I set an environment variable in hubot.conf? What's a secure method?

**Comment by jfryman** [comment]
The best way to sent environment variables is either in the process supervisor that you are using to manage Hubot, or the entry script (`bin/hubot`). In there, you'll need to set a few env vars (see https://github.com/stackstorm/hubot-stackstorm#plugin-environment-variable-options) and you'll be good to go!

**Comment by Kami** [comment]
It looks like this is the same issue as #45.

I will make sure this "limitation" gets noted in the readme for now.

**Comment by emedvedev** [comment]
This one was fixed some time ago.

**Comment by Sphinxs** [comment]
Why would I receive this error using just the Stack Storm in a Docker container and following the quick tutorial, I have logged in at the web interface and in the CLI...?

**Comment by blag** [comment]
@Sphinxs Can you point me to the quick tutorial you used? There are a few "quick tutorials" of StackStorm around.

If you are running ST2 as one-microservice-per-container, then I suspect that your st2chatops container is not configured properly. Check the `ST2_API_KEY` environment variable in the container and make sure it is a valid API key for the ST2 API, or check that the `ST2_API_KEY` in `/opt/stackstorm/chatops/st2chatops.env` is set to a valid value.

If you are running all ST2 microservices in a single container, then check that the st2chatops microservice is configured correctly in `/opt/stackstorm/chatops/st2chatops.env`.

**Comment by Sphinxs** [comment]
The tutorial was [this](https://docs.stackstorm.com/start.html) one, I'm going to check. I'm running as a single container.

**Comment by Sphinxs** [comment]
I get how it works watching the [StackStorm 101 (v0.12)](https://www.youtube.com/watch?v=CwEzO_f-q3s)

**Comment by blag** [comment]
Okay, just keep in mind that that tutorial is over four years old at this point, and is running StackStorm 0.12. The last released version of StackStorm was 3.1, so there's been quite a few changes since then and things may not work the same way. And hubot-stackstorm in particular has seen some [huge refactoring](https://github.com/StackStorm/hubot-stackstorm/pulls?q=is%3Apr+is%3Aclosed+label%3Arefactor) lately.

I don't really know how one would (properly) run Atom in a container, and I'm pretty sure that would cause further issues for you. Either way, it's outside the scope of this issue, but if you really would like to find the answer, you can try asking in Docker's own [community](https://www.docker.com/docker-community).

Running StackStorm in Docker is difficult, so for pack development, I would highly recommend running it in a Vagrant container using this repository:
https://github.com/StackStorm/st2vagrant

That will remove one large layer of configuration complication (eg: Docker) for you, just make sure you read and follow the README for that repository. It will install and configure ST2 for you, although you will need to add your Slack token to `/opt/stackstorm/chatops/st2chatops.env` and restart chatops (`st2ctl restart-component st2chatops`). Running ST2 in a VM is really how ST2 was designed to be run, even though we have been working towards running it in Docker and Kubernetes, so the more you stick with that the fewer confusing issues you will have.

The README for st2vagrant also has directions for how to configure a shared volume between the host and the ST2 VM, so you can modify the pack on your host (in Atom!) and test it out in the VM.

Our documentation on packs themselves is here:
https://docs.stackstorm.com/packs.html

And our guide to creating and contributing packs is here:
https://docs.stackstorm.com/reference/packs.html

You can also search for free, pre-existing packs on StackStorm Exchange:
https://exchange.stackstorm.org

Hopefully some of those are useful to you so you don't have to reinvent any wheels. :smile:

Lastly, instead of necromancing a long-resolved GitHub issue, you may find better support by asking in our [forum](https://forum.stackstorm.com), or especially in our [Slack community](https://stackstorm-community.slack.com). If you need more individualized support, you can contact Extreme Sales by filling out the form on this page:
https://www.extremenetworks.com/contact-sales
